RE: Consolidate API requests for the home page shelves

xlisachan opened the pull request:
Fixes #9996 

This pull request is a work-in-progress to utilize the homepage shelves API.

@bobsilverberg - as discussed, I am still working on the footer portion (text and pathname) of the component, as well as correcting the flow error in the .map method. I also believe I need to include `i18n` for the header and footer text, correct?

While I was able to remove collections, heroShelves, and shelves from state, I am having trouble removing collections and shelves completely from the reducers and saga. Any insight would be greatly appreciated! :)

Please review when you have a moment. Thank you!

bobsilverberg reviewed (commented):
I think you've already addressed the one comment I added, and I don't see anything else that jumps out at me. Please let me know once you have another version for me to look at.

bobsilverberg commented:
@xlisachan Our discussion about generating the `footer_text` was incorrect. If all we need to do is concatenate some static text with some text from the database, we can do it via `i18n.sprintf`. See https://github.com/mozilla/addons-frontend/blob/master/src/amo/pages/Addon/index.js#L160-L162 for an example.

xlisachan commented:
@bobsilverberg, restarting my computer fixed the console issues.

For `amo/reducers/versions.js`
`homeShelves`, `homeShelves.results`, `Object.keys(shelves)` all consoles what was expected.
<img width="1322" alt="Screen Shot 2021-02-23 at 1 03 25 PM" src="https://user-images.githubusercontent.com/28129806/108887125-d19bd000-75d7-11eb-9067-7122e0f1a9d3.png">
<img width="369" alt="Screen Shot 2021-02-23 at 1 53 18 PM" src="https://user-images.githubusercontent.com/28129806/108893061-a2d52800-75de-11eb-9ca5-bdcb4fb4bdb3.png">

However, if I try to go further, I get the error `In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`
<img width="998" alt="Screen Shot 2021-02-23 at 1 58 11 PM" src="https://user-images.githubusercontent.com/28129806/108893609-4a525a80-75df-11eb-9fb4-9b5a6308d6ed.png">

The other issue is when I run tests and repeatedly get a type error - `Cannot read property 'type' of undefined` for `createInternalShelf` in `amo/reducers/home.js`.
<img width="1122" alt="Screen Shot 2021-02-23 at 2 01 32 PM" src="https://user-images.githubusercontent.com/28129806/108894178-fa27c800-75df-11eb-9f9d-51fa227183a8.png">

Please review when you have a moment. Thank you!!

bobsilverberg commented:
@xlisachan Do you think you could rebase this PR against the current master branch? It's not letting me pull it down as is. In the meantime I'll see if I can figure out any of your issues.

bobsilverberg commented:
> @xlisachan Do you think you could rebase this PR against the current master branch? It's not letting me pull it down as is. In the meantime I'll see if I can figure out any of your issues.

I was able to get around this issue, but ultimately you'll need to rebase it, so now is as good a time as any to do that. Thanks!

bobsilverberg commented on src/amo/reducers/versions.js:
```diff
@@ -500,42 +500,47 @@ const reducer = (
+      //   }
+      // }
+
+      for (const shelves of homeShelves.results) {
+        // eslint-disable-next-line no-console
+        console.log(shelves);
       }
```
```suggestion
     for (const shelf of homeShelves.results) {
        for (const addon of shelf.addons) {
          if (addon.current_version) {
            const currentVersion = addon.current_version;
            let version = createInternalVersion(currentVersion, state.lang);
            // license and release_notes are omitted from the search endpoint results,
            // use them from an existing version if available.
            const { id } = currentVersion;
            const existingVersion = getVersionById({ id, state });
            if (existingVersion) {
              version = {
                ...version,
                license: existingVersion.license,
                releaseNotes: existingVersion.releaseNotes,
              };
            }
            newVersions[id] = version;
          }
        }
      }

```

bobsilverberg replied:
I believe this code will work to address the issue you were having in the versions reducer.

bobsilverberg reviewed (changes requested):
@xlisachan I made a suggestion which should fix your reducer issue. I will look at the tests in a bit as well.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,73 @@
+
+    const homepageShelves = shelves.map((shelf) => {
+      const { addons, endpoint, footer_text, title } = shelf;
+      const header = title.replace(/\s/g, '');
```
After fixing the problem with the `versions` reducer, I noticed this is now causing an error, because your code is assuming `title` will be a string, but it's actually now a localized string (i.e., an object) returned from the API. You'll need to deal with these now localized fields in the routine where you create the internal representation of the shelves, and that should fix this line without requiring any changes to this file.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -249,11 +252,28 @@ type Action =
+  });
+
+  return {
+    title: result.title,
```
Here's one of the spots where you'll need to deal with the fact that the API is now returning localized strings. There are others as well.

bobsilverberg commented on src/amo/pages/Home/index.js:
```diff
@@ -43,18 +37,6 @@ export const FEATURED_COLLECTIONS = [
   { slug: 'password-managers', userId: MOZILLA_USER_ID },
 ];
 
-export const isFeaturedCollection = (
```
This function was also used in `amo/pages/Collection/index.js`, but I don't think it was actually needed there, so its usage from that file can be removed as well. Without doing that we get test failures.

bobsilverberg commented on tests/unit/helpers.js:
```diff
@@ -295,6 +295,28 @@ export const fakePrimaryHeroShelfExternalAddon = Object.freeze({
+  criteria = '?sort=rating&type=statictheme',
+  footerText = 'See more top rated themes',
+  footerPathname = 'search/?sort=rating&type=statictheme',
+  addons = fakeAddon,
```
```suggestion
  addons = [fakeAddon],
```

bobsilverberg replied:
This is part of what's needed to fix those test errors about "Cannot read property 'type' of undefined".

bobsilverberg commented on tests/unit/helpers.js:
```diff
@@ -378,11 +400,13 @@ export const createSecondaryHeroShelf = ({
 
-export const createHeroShelves = ({
+export const createHomeShelves = ({
+  resultsProps = [],
```
```suggestion
  results,
```

bobsilverberg replied:
Another change to fix the test failures. There's one more...

bobsilverberg commented on tests/unit/helpers.js:
```diff
@@ -378,11 +400,13 @@ export const createSecondaryHeroShelf = ({
   secondaryProps = {},
 } = {}) => {
   return {
+    results: createResultsShelf(resultsProps),
```
```suggestion
    results: results || createResultsShelf(),
```

bobsilverberg replied:
Applying these three suggestions to this file will fix those errors. It changes the interface for `createHomeShelves`, but I think that's fine.

bobsilverberg reviewed (changes requested):
@xlisachan I added a few more comments, including some code suggestions for fixing the errors you were getting from some tests. There are still quite a lot of failures for the tests for the Home page, as well as the `home` and `versions` reducers. I can jump in and help with some, or all, of those if you like. Just let me know.

xlisachan commented:
@bobsilverberg, the latest pull request includes updates to the home/collections pages (and tests), the home and versions reducers (and tests), home saga, and helpers. There are a few tests that I was having trouble including, please see below for where I left off on each test for you to take over as discussed.
(1) `tests/unit/amo/pages/TestHome.js`
`does not fetch the add-ons when results are already loaded` - I think this test should be updated/moved to the HomepageShelves component test?

(2) `tests/unit/amo/reducers/test_home.js`
`loads the correct amount of theme add-ons in a collection to display on homepage` - I tried updating this to '... in a 
 collection homepage shelf', but kept getting a `Cannot read property 'type' of undefined` error. I also think this should also be moved to the HomepageShelves component test as well though?

(3) `tests/unit/amo/reducers/test_version.js`
I tried to add `loads versions for homeShelves`, but kept getting mismatched received/expected arrays.

I am currently updating the HomepageShelves component and its test, and will be pushing those through as well.

Please review when you have a moment. Thank you!!

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+import * as React from 'react';
+import config from 'config';
+
+import translate from 'amo/i18n/translate';
```
Nit: We try to keep these in alphabetical order, based on whence the import is from, so the import from `amo/i18n/translate` would come after the import from `amo/constants`, because `constants` is before `i18n` in the alphabet.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+export const MOZILLA_USER_ID = config.get('mozillaUserId');
+
+type Props = {|
+  i18n: Object,
```
Because `i18n` isn't a prop that is passed directly to this component from a parent (but rather included via `translate()`), it should go into a type called `InternalProps`. See other components for an example of this. It also has its own type.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+      const { addons, criteria, endpoint, footer, title } = shelf;
+      const titleStr = title.toString();
+      const header = titleStr.replace(/\s/g, '');
+      const footerText = footer ? footer.text : 'See more';
```
We need to use `i18n.sprintf` along with `i18n.gettext` to create a translatable template for this, where we can substitute in the shelf title. You can find lots of examples of this in other components.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+
+      const className =
+        endpoint === 'collections'
+          ? 'Home-FeaturedCollection'
```
This was just used for testing purposes, so I don't think we need it anymore. If every shelf ends up with a className like `Home-${header}` that should suffice for testing.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+    const homepageShelves: Array<React.Node> = shelves.map((shelf) => {
+      const { addons, criteria, endpoint, footer, title } = shelf;
+      const titleStr = title.toString();
+      const header = titleStr.replace(/\s/g, '');
```
It might make it more readible if you were to add a dash (`-`) instead of an empty string.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+          addonInstallSource={addonInstallSource}
+          addons={addons}
+          className={className}
+          footerText={i18n.sprintf(i18n.gettext('%(text)s'), {
```
Oh, I see you're using `i18n.sprintf` here. Does this actually work? This substitution doesn't look right to me.

xlisachan replied:
It works - it looked odd to me as well though. 

I had tried `{i18n.sprintf(i18n.gettext(footerText))}`, but the error stated that `i18n.sprintf` requires another argument. 
In addition, when I tried `{i18n.sprintf(i18n.gettext('%s'), footerText)}`, the error stated the second parameter of `i18n.sprintf` was incompatible with object type.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -146,26 +141,52 @@ export type HeroShelvesType = {|
+|};
+
+export type HomeShelvesType = {|
+  results: Array<ResultShelfType> | null,
```
When would this be `null`?

bobsilverberg replied:
This should also never be `null`.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -293,11 +283,36 @@ export const createInternalSecondaryHeroModule = (
+  }
 
   let secondaryShelf: SecondaryHeroShelfType | null = null;
   if (secondary !== null) {
```
Same here, when would `secondary` be `null`?

xlisachan replied:
Lines 302-315 were added by Andrew for the update to handle when the API returned localized fields from the API...

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -314,7 +329,11 @@ export const createInternalHeroShelves = (
   }
 
   if (primary === null) {
```
Same here, when would `primary` be `null`?

xlisachan replied:
This was a part of the original `createInternalHeroShelves` that was updated to handle cases in which there was missing hero data.

bobsilverberg commented on src/amo/sagas/home.js:
```diff
@@ -1,184 +1,32 @@
 
-  const state = yield select(getState);
-
-  const recommendedSearchFilters = {
```
As discussed above, we're either going to have to enhance the curation tool to allow us to specify different pages for desktop vs. mobile, or we're going to have to keep the code here which retrieves just the Recommended Extensions when we are working with the mobile page. My feeling is that we should do the latter for now, and treat the changes to the curation tool as a possible follow up.

bobsilverberg reviewed (changes requested):
This is great work, thanks Lisa! It is coming along very well. I added quite a few comments, but most are pretty minor. Please let me know if you have any questions about any of them. I guess we can discuss them at greater length in tomorrow's meeting.

I didn't review your changes to the tests yet, as I believe our plan is for me to take over work on the tests.

xlisachan commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,82 @@
+          })}
+          footerLink={footerLink}
+          header={titleStr}
+          isTheme={endpoint === 'search-themes' ? true : undefined}
```
I believe `false` had not worked, but this was intended for an earlier edit with `FeaturedCollectionCard` before I realized that that component utilized `LandingAddonsCard` as well.

I have updated it to `isTheme={endpoint === 'search-themes'}` as advised.

xlisachan commented:
@bobsilverberg , I just wanted to provide an update -

I pushed through revisions to the following component and page:
- `amo/components/HomepageShelves/index.js`
- `amo/pages/Collection/index.js`

I pushed through a WIP for `amo/sagas/home.js`, as I have a question. `recommendedExtensions` returns just the array of addons from its API. As `homeShelves.results` is an array of shelves, we do not have the shelf detail for Recommended Extensions. Should we proceed with hard-coding these details? If so, I am not sure what to enter for `url` and the footer object.

I am still working through the revisions for the following: 
- `amo/pages/Home/index.js`
- `amo/reducers/home.js`

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,93 @@
+      const footerText =
+        footer && footer.text
+          ? footer.text
+          : i18n.sprintf(i18n.gettext('See more %(text)s'), {
```
Just to make this a bit more readable, maybe change `text` to `categoryName`.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -146,26 +141,52 @@ export type HeroShelvesType = {|
+|};
+
+export type ResultShelfType = {|
+  title: LocalizedString,
```
Just pointing out that this has yet to be addressed.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -146,26 +141,52 @@ export type HeroShelvesType = {|
+|};
+
+export type ExternalHomeShelvesType = {|
+  results: Array<ExternalResultShelfType> | null,
```
We discussed changing this to assume it will never be `null`, and an issue was filed for addons-server to change the API so it never can be `null`.

bobsilverberg reviewed (commented):
> @bobsilverberg , I just wanted to provide an update -
> 
> I pushed through revisions to the following component and page:
> 
> * `amo/components/HomepageShelves/index.js`
> 
> * `amo/pages/Collection/index.js`
> 
> 
> I pushed through a WIP for `amo/sagas/home.js`, as I have a question. `recommendedExtensions` returns just the array of addons from its API. As `homeShelves.results` is an array of shelves, we do not have the shelf detail for Recommended Extensions. Should we proceed with hard-coding these details? If so, I am not sure what to enter for `url` and the footer object.
>

Thanks for the updates @xlisachan, it's looking good!
 
I think the approach you took with hardcoding is good for now. When we get closer to a finished product we might decide to move that elsewhere, but for now it's good. As for the URLs, I'm not sure. I'll finish reviewing the latest version of your patch and then I'll see if I have any suggestions.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -226,19 +230,12 @@ type LoadHomeDataAction = {|
+  homeShelves,
 }: LoadHomeDataParams): LoadHomeDataAction => {
-  invariant(collections, 'collections is required');
-  invariant(shelves, 'shelves is required');
```
I think we should keep an `invariant` for `homeShelves`. Is there ever a time where we'd be calling `loadHomeData` and `homeShelves` is falsey?

bobsilverberg commented on src/amo/sagas/home.js:
```diff
@@ -60,121 +53,57 @@ export function* fetchHomeData({
+    let homeShelves = {};
     try {
-      heroShelves = yield call(getHeroShelves, { api: state.api });
+      homeShelves = yield call(getHomeShelves, { api: state.api });
```
I don't think we need the hero shelves data when we are on the mobile page, so we can probably avoid calling this API altogether if `isDesktopSite` is `false`. In that case we only need to call the API to retrieve the Recommended Extensions shelf.

bobsilverberg commented on src/amo/sagas/home.js:
```diff
@@ -60,121 +53,57 @@ export function* fetchHomeData({
+    addons: Array<ExternalAddonType>,
+  ): ExternalResultShelfType => {
+    return {
+      title: selectLocalizedContent('Recommended Extensions', state.lang),
```
What you are creating here is meant to be an external representation of the data (i.e., what we would get back from an API), so you don't want to use `selectLocalizedContent` here, as that's used for creating an internal representation. I think what we actually want here is `createLocalizedString`. The problem with that is that it lives in `tests/unit/helpers.js` and we can't really use that in our code, so it will need to be moved.

I think we can make that a follow-up, so for now you should just contruct a LocalizedString object in a way similar to what `createLocalizedString` does here. Because you'll need to do it twice (once here, and once for `footer.text`, you should create a local helper function for it.

bobsilverberg reviewed (changes requested):
Thanks for the updates @xlisachan. This is really great work! I can do all of the work on the tests, but I think it makes sense for me to wait for you to address the new comments first, to avoid having to write the tests more than once. Please let me know if you have any questions about any of my comments.

xlisachan commented:
Thanks for the feedback, @bobsilverberg!

I have pushed through revisions for the following:
- `amo/components/HomepageShelves/index.js`
     - addresses change of InternalProps to Props
     - readability with the `footerText` (`text` to `categoryName`)
     - fixes footer's url, and
     - adds in the loading skeleton when `loading` is true

- `amo/pages/Home/index.js`
     - revised so that `HomepageShelves` component is always included in Home if isDesktopSite is true, otherwise it will display the Recommended Extensions shelf via `LandingAddonsCard`
     
- `amo/reducers/home.js`
     - changed `LocalizedString` to `string` for `title` of `ExternalResultShelfType` and `ResultShelfType`
     - changed `null` to `[]` for `results` in `ExternalHomeShelvesType` and `HomeShelvesType`,
     - added `invariant` for `homeShelves`, and
     - added `shelves` data to handle mobile homepage logic
   
- `amo/sagas/home.js`
    - Moved to call homeShelves API for when only isDesktopSite is true 
  
- `amo/reducers/versions.js
    - Added back in shelves data

I removed creating the Recommended Extensions shelf in `amo/sagas/home.js` and opted to utilize the `LandingAddonsCard` for the mobile homepage logic, as it handles the concerns for the footer url until it is resolved.

Please review when you have a moment. Thank you!

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,93 @@
+
+export default (translate()(
+  HomepageShelves,
+): React.ComponentType<InternalProps>);
```
It doesn't look like this has been updated.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+
+    const MOZILLA_USER_ID = config.get('mozillaUserId');
+
+    const homepageShelves: Array<React.Node> = shelves.map((shelf) => {
```
The issue with this approach is that, even if `loading` is `true`, we are still doing some work to create `homepageShelves`. A different approach might be to have a variable called `shelvesContent` and have an `if` statement at the top setting `shelvesContent` to one thing is `loading` is `true` and to the other thing if `loading` is `false. That way you'll only build the shelf content you need in each case.

xlisachan replied:
I restructured the component and then tried to update it based on your comments, but I am not entirely sure if it's what you meant... please take a look.

bobsilverberg replied:
It's not exactly what I meant. I'll make a diff to show you what I am suggesting.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+        if (footer.url.startsWith('/')) {
+          footerUrl = footer.url;
+        } else {
+          footerUrl = `/${footer.url}`;
```
Just wondering where this requirement comes from? What if it's a fully qualified URL (e.g., `https://addons.mozilla.org/en-US/firefox/collections/4757633/tab-organizers`)?

xlisachan replied:
I had thought `footer.url` would be a relative path and therefore received the error - `Error: "to" prop cannot contain a relative path; it must start with a "/".` I have updated it to expect admin to always make `footer.url` a fully qualified URL, please let me know your thoughts.

bobsilverberg replied:
I think we need to just assume that whatever they put is accurate, because they can put anything they want here. It should generally be an internal URL, but could be an external one as well, theoretically.

bobsilverberg replied:
Actually, looking at the code you have to process this, this gets a bit more complicated. Let me check to see what might be used in there.

bobsilverberg replied:
We can support only internal URLs, so I think the code you have here now is good.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -226,20 +238,14 @@ type LoadHomeDataAction = {|
-  collections,
-  heroShelves,
+  homeShelves,
   shelves,
```
I'm not really a huge fan of keeping `shelves` around, but I can see why you're doing it. I suppose it is the lesser evil, so let's leave it for now, unless I can think of a different, better way.

bobsilverberg reviewed (changes requested):
Thanks for the updates @xlisachan, it's looking really good! I think we're close, and I'll start working on the tests in earnest. I've added some comments for you to address.

xlisachan commented:
Thanks for the feedback, @bobsilverberg! The latest push includes updates based on your comments today.

As mentioned above, I just wanted to make sure I interpreted the `shelvesContent` revision correctly in my update. In addition for the `footer.url`, I updated it to expect admin to always make `footer.url` a fully qualified URL, please let me know if I should not make this assumption.

Please review when you have a moment. Thank you!!

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+
+    let footerLinkHtml;
+    if (footer && footer.url) {
+      footerLinkHtml = <a href={footer.url}>{footerText}</a>;
```
This isn't right, we shouldn't have to put html here. I can see from `LandingAddonsCard` that it expects `footerLink` to be an internal URL, so if we want to support an external URL we'll need to change that as well. I have asked a question in the issue to see if we will need to support external URLs.

bobsilverberg replied:
See above, we can assume it will always be an internal URL and just pass it into the `to`.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+class HomepageShelves extends React.Component<Props> {
+  /* eslint-disable react/no-array-index-key */
+  static defaultProps: DefaultProps = {
+    placeholderCount: LANDING_PAGE_EXTENSION_COUNT,
```
I know I've reviewed this a few times now and this hasn't come up, but I just found myself wondering, why do we need this prop? This component will only ever be called from one place (the home page), so why does this need to be a prop? Could we just define this in here as a constant and be done with it?

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+|};
+
+class HomepageShelves extends React.Component<Props> {
+  /* eslint-disable react/no-array-index-key */
```
We shouldn't need this. I can see it's needed for the key for the loading state shelves, but you can do that another way. I'll add a suggestion to that.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+    placeholderCount: LANDING_PAGE_EXTENSION_COUNT,
+  };
+
+  homepageShelves: Array<React.Node> = this.props.shelves.map((shelf) => {
```
Here is a diff showing the change I was suggesting:

```diff
diff --git a/src/amo/components/HomepageShelves/index.js b/src/amo/components/HomepageShelves/index.js
index 1104c8020..871bfff75 100644
--- a/src/amo/components/HomepageShelves/index.js
+++ b/src/amo/components/HomepageShelves/index.js
@@ -38,86 +38,82 @@ class HomepageShelves extends React.Component<Props> {
     placeholderCount: LANDING_PAGE_EXTENSION_COUNT,
   };

-  homepageShelves: Array<React.Node> = this.props.shelves.map((shelf) => {
-    const { addons, criteria, endpoint, footer, title } = shelf;
-    const MOZILLA_USER_ID = config.get('mozillaUserId');
-    const titleStr = title.toString();
-    const header = titleStr.replace(/\s/g, '-');
-
-    const footerText =
-      footer && footer.text
-        ? footer.text
-        : this.props.i18n.sprintf(
-            this.props.i18n.gettext('See more %(categoryName)s'),
-            {
-              categoryName: titleStr.toLowerCase(),
-            },
-          );
-
-    const addonInstallSource =
-      endpoint === 'collections'
-        ? INSTALL_SOURCE_FEATURED_COLLECTION
-        : INSTALL_SOURCE_FEATURED;
-
-    const count =
-      endpoint === 'search-themes'
-        ? LANDING_PAGE_THEME_COUNT
-        : this.props.placeholderCount;
-
-    let footerLinkHtml;
-    if (footer && footer.url) {
-      footerLinkHtml = <a href={footer.url}>{footerText}</a>;
-    }
-
-    const defaultFooterUrl =
-      endpoint === 'collections'
-        ? `/collections/${MOZILLA_USER_ID}/${criteria}/`
-        : `/search/${criteria}/`;
-
-    const footerLink = footerLinkHtml || defaultFooterUrl;
-
-    return (
-      <LandingAddonsCard
-        addonInstallSource={addonInstallSource}
-        addons={addons}
-        className={`Home-${header}`}
-        footerText={footerText}
-        footerLink={footerLink}
-        header={titleStr}
-        isTheme={endpoint === 'search-themes'}
-        key={header}
-        loading={this.props.loading}
-        placeholderCount={count}
-      />
-    );
-  });
-
-  loadingShelves = (
-    <div className="HomepageShelves-loading">
-      {Array(3)
-        .fill(0)
-        .map((value, index) => {
-          return (
-            <LandingAddonsCard
-              className="HomepageShelves-loading-text"
-              key={`HomepageShelves-loading-text-${index}`}
-              header={<LoadingText width={100} />}
-              loading={this.props.loading}
-            />
-          );
-        })}
-    </div>
-  );
-
   render(): React.Node {
-    const { loading } = this.props;
+    const { loading, shelves } = this.props;

     let shelvesContent;

-    if (loading === true) {
-      shelvesContent = this.loadingShelves;
+    if (loading) {
+      shelvesContent = (
+        <div className="HomepageShelves-loading">
+          {Array(3)
+            .fill(0)
+            .map((value, index) => {
+              return (
+                <LandingAddonsCard
+                  className="HomepageShelves-loading-text"
+                  key={`HomepageShelves-loading-text-${index}`}
+                  header={<LoadingText width={100} />}
+                  loading={this.props.loading}
+                />
+              );
+            })}
+        </div>
+      );
     } else {
-      shelvesContent = this.homepageShelves;
+      shelvesContent = shelves.map((shelf) => {
+        const { addons, criteria, endpoint, footer, title } = shelf;
+        const MOZILLA_USER_ID = config.get('mozillaUserId');
+        const titleStr = title.toString();
+        const header = titleStr.replace(/\s/g, '-');
+
+        const footerText =
+          footer && footer.text
+            ? footer.text
+            : this.props.i18n.sprintf(
+                this.props.i18n.gettext('See more %(categoryName)s'),
+                {
+                  categoryName: titleStr.toLowerCase(),
+                },
+              );
+
+        const addonInstallSource =
+          endpoint === 'collections'
+            ? INSTALL_SOURCE_FEATURED_COLLECTION
+            : INSTALL_SOURCE_FEATURED;
+
+        const count =
+          endpoint === 'search-themes'
+            ? LANDING_PAGE_THEME_COUNT
+            : this.props.placeholderCount;
+
+        let footerLinkHtml;
+        if (footer && footer.url) {
+          footerLinkHtml = <a href={footer.url}>{footerText}</a>;
+        }
+
+        const defaultFooterUrl =
+          endpoint === 'collections'
+            ? `/collections/${MOZILLA_USER_ID}/${criteria}/`
+            : `/search/${criteria}/`;
+
+        const footerLink = footerLinkHtml || defaultFooterUrl;
+
+        return (
+          <LandingAddonsCard
+            addonInstallSource={addonInstallSource}
+            addons={addons}
+            className={`Home-${header}`}
+            footerText={footerText}
+            footerLink={footerLink}
+            header={titleStr}
+            isTheme={endpoint === 'search-themes'}
+            key={header}
+            loading={this.props.loading}
+            placeholderCount={count}
+          />
+        );
+      });
     }

     return <div className="Home-HomepageShelves">{shelvesContent}</div>;

```

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+
+  loadingShelves = (
+    <div className="HomepageShelves-loading">
+      {Array(3)
```
To fix the problem you were having with `react/no-array-index-key`, you could replace:

`{Array(3).fill(0).map((value, index) => {`

with 

`{[1,2,3].map((key) => {`

and then use `key` in place of `index` below on line 103.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+            <LandingAddonsCard
+              className="HomepageShelves-loading-text"
+              key={`HomepageShelves-loading-text-${index}`}
+              header={<LoadingText width={100} />}
```
Does it need this `header`? I would have thought that just specifying `loading` would be enough to make it look correct.

xlisachan replied:
Without including `header`, it's missing where the title of the shelf goes in the skeleton (see examples below). Please let me know which is preferred.

With header:
<img width="980" alt="Screen Shot 2021-03-16 at 6 05 16 PM" src="https://user-images.githubusercontent.com/28129806/111410350-f2bf8000-86ae-11eb-868b-fce89e425cc0.png">

Without header:
<img width="980" alt="Screen Shot 2021-03-16 at 6 04 24 PM" src="https://user-images.githubusercontent.com/28129806/111410366-f94df780-86ae-11eb-83eb-47455f585c17.png">

bobsilverberg replied:
Cool, thanks. It looks like we do need it.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,127 @@
+              className="HomepageShelves-loading-text"
+              key={`HomepageShelves-loading-text-${index}`}
+              header={<LoadingText width={100} />}
+              loading={this.props.loading}
```
This only happens when `loading` is `true`, so instead of `loading={this.props.loading}` you can just have `loading`.

bobsilverberg commented on src/amo/components/HomepageShelves/styles.scss:
```diff
@@ -0,0 +1,6 @@
+@import '~amo/css/styles';
+
+.HomepageShelves-loading-text {
```
I'm surprised you need this. I would have thought it would have "just worked", as it works on the current homepage.

xlisachan replied:
That was an oversight, sorry! It does work without the import.

bobsilverberg replied:
I'm not sure what you mean by "the import".

xlisachan replied:
Ohhh, I thought you meant that I didn't need `@import '~amo/css/styles';` for it to work. I did not realize that I didn't need the stylesheet at all; I'll delete the stylesheet. 🤦🏻‍♀️

bobsilverberg commented on src/amo/pages/Home/index.js:
```diff
@@ -275,120 +193,47 @@ export class HomeBase extends React.Component {
+            {isDesktopSite ? (
+              <HomepageShelves
+                loading={loading}
+                placeholderCount={LANDING_PAGE_EXTENSION_COUNT}
```
It seems that we always pass `LANDING_PAGE_EXTENSION_COUNT` into `placeholderCount`, so we might as well dispense with that prop and just define it in `HomepageShelves`. I mentioned this above as well.

bobsilverberg commented on src/amo/reducers/versions.js:
```diff
@@ -472,14 +472,25 @@ const reducer = (
-            );
-            newVersions[version.id] = version;
+      for (const shelf of Object.keys(shelves)) {
+        if (shelves[shelf]) {
```
This code in here looks like it's duplicated between here and the code that processes the `homeShelves`. It should be extracted in a way that it can be reused in both places.

bobsilverberg replied:
Nice work extracting this code!

bobsilverberg reviewed (changes requested):
It's looking really good, thanks for all your efforts @xlisachan! I've added some more comments.

xlisachan commented:
Thanks again for the feedback, @bobsilverberg ! I made the requested updates. In addition -
- I'll wait for further insight regarding the footer.url.
- Please let me know if you prefer the homepage shelves loading skeleton with or without the header (see screenshots above)

Please review when you have a moment. Thank you!

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,107 @@
+
+        let footerLinkHtml;
+        if (footer && footer.url) {
+          footerLinkHtml = <a href={footer.url}>{footerText}</a>;
```
I replied to some earlier comments, but I'll add one here as well. If we have a `footer.url` we can just use it as is, although the code you added to add a leading slash is a good idea, but we never need to wrap it in an `<a>` tag.

bobsilverberg commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,107 @@
+            header={titleStr}
+            isTheme={endpoint === 'search-themes'}
+            key={header}
+            loading={loading}
```
As above, `loading` will always be `false` here, so you can just leave this prop out.

xlisachan replied:
If I remove the line `loading={loading}`, I get a flow error saying `Cannot create LandingAddonsCard element because property 'loading' is missing in props but exists in Props.` Should it be left as is or updated to `loading={false}`?

bobsilverberg replied:
We should update `LandingAddonsCard` to make `loading` optional, which would address that issue. That would mean changing the type, and perhaps some code, but maybe just the type, as it's a boolean and therefore `undefined` should work just as well as `false`. Do you want to give that a try?

bobsilverberg reviewed (changes requested):
Thanks again for the updates @xlisachan, great work! I responded to your questions about the loading header and the footer URL. I will continue working on the tests.

xlisachan commented:
Thanks for the feedback, @bobsilverberg ! The latest push includes the following updates:
- Removing HomepageShelves stylesheet and reference in the `HomepageShelves` component
- Updating `footer.url` logic to use if provided (and checking it starts with '/')
- Revising `loading` type to optional in `LandingAddonsCard` and removing `loading` props when it's false in the `HomepageShelves` component

Please review when you have a moment. Thank you!

bobsilverberg commented:
> Thanks for the feedback, @bobsilverberg ! The latest push includes the following updates:
> 
> * Removing HomepageShelves stylesheet and reference in the `HomepageShelves` component
> 
> * Updating `footer.url` logic to use if provided (and checking it starts with '/')
> 
> * Revising `loading` type to optional in `LandingAddonsCard` and removing `loading` props when it's false in the `HomepageShelves` component
> 
> 
> Please review when you have a moment. Thank you!

Thanks @xlisachan. You're moving so quickly I can't keep up with you! I'm working on the tests right now, and I've made a few minor changes to some code along the way. My thought is that I can make a PR to your branch once I have the tests done, and then you can pull them in and push the whole thing back to here. I probably won't have anything for you until tomorrow though.

bobsilverberg commented:
I'll push whatever I have at the end of the day so that should at least address the tests that are currently failing.

bobsilverberg commented on src/amo/pages/Home/index.js:
```diff
@@ -194,28 +137,24 @@ export class HomeBase extends React.Component {
   }
 
   renderHeroHeader() {
```
I made a mistake with this thing. This is included on mobile, which means that we need `homeShelves` on mobile as well as on desktop. This will change only the saga, I think, but we need to make sure that we call the API and load the data from the shelves API, even on mobile.

xlisachan commented:
No worries! I revised saga to include `homeShelves` on mobile as well. :)

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -146,26 +141,52 @@ export type HeroShelvesType = {|
+  url: string,
+  endpoint: string,
+  criteria: string,
+  footer: HeroCallToActionType | null,
```
I addressed this in my latest commit.

bobsilverberg commented:
I think this might be ready to do, @xlisachan. Amazing work on your part! I'm going to look over it one more time, and then request review from another member of the team, as I did some work on the tests. With any luck it will be ready to land next week!

bobsilverberg reviewed (approved):
It all looks good to me now, thanks @xlisachan! I'll request a review from one of my teammates.

xlisachan commented:
Thanks for your guidance, @bobsilverberg ! And thank you for working on the tests! :)

willdurand commented on tests/unit/helpers.js:
```diff
@@ -378,11 +392,13 @@ export const createSecondaryHeroShelf = ({
 
-export const createHeroShelves = ({
+export const createHomeShelves = ({
+  resultsProps,
```
could we set a default value to this param instead of having a fallback later? Maybe something like this?

```suggestion
  resultsProps = [fakeShelf],
```

willdurand commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -52,159 +43,55 @@ describe(__filename, () => {
     sagaTester.start(homeSaga);
+
+    const state = sagaTester.getState();
+    baseArgs = { api: state.api };
```
Maybe:

```suggestion
    baseApiArgs = { api: state.api };
```

`baseArgs` is a bit too generic.

willdurand commented on src/amo/sagas/home.js:
```diff
@@ -1,179 +1,76 @@
       );
     } else {
-      // Mobile homepage logic.
+      // Mobile homepage logic
```
🐼

willdurand commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,119 @@
+    });
+  }
+
+  return <div className="Home-HomepageShelves">{shelvesContent}</div>;
```
We usually set the top-level class name to the name of the component, so it would be `HomepageShelves` in this case.

willdurand commented on src/amo/components/HomepageShelves/index.js:
```diff
@@ -0,0 +1,119 @@
+  if (loading) {
+    shelvesContent = (
+      <div className="HomepageShelves-loading">
+        {[1, 2, 3].map((key) => {
```
Add a comment to explain why we display 3 loading shelves.

willdurand reviewed (commented):
I don't have many comments on the code itself, there have been enough iterations.

You should update the PR description, and make sure to include "Fixes #9996".

> Merging [#10065](https://codecov.io/gh/mozilla/addons-frontend/pull/10065?src=pr&el=desc) ([57b27cb](https://github.com/mozilla/addons-frontend/commit/57b27cb31c2c2712059c189e1e492ed83910da61)) into [master](https://codecov.io/gh/mozilla/addons-frontend/commit/c58c53ba1e0117b6a28381d8837e86826ebb1acb?el=desc) ([c58c53b](https://github.com/mozilla/addons-frontend/commit/c58c53ba1e0117b6a28381d8837e86826ebb1acb)) will **decrease** coverage by `0.37%`.
> The diff coverage is `63.79%`.

mm this is weird. The diff coverage remains 63% but when looking in codecov, it seems a bit better than that. We definitely want 100% diff coverage.

I'd suggest you rebase your PR, too.

bobsilverberg commented:
Thanks for the review @willdurand!  @xlisachan Can you take a look and see if you feel you can address these suggestions? Let me know if you need any help.

> I don't have many comments on the code itself, there have been enough iterations.
> 
> You should update the PR description, and make sure to include "Fixes #9996".
> 
> > Merging [#10065](https://codecov.io/gh/mozilla/addons-frontend/pull/10065?src=pr&el=desc) ([57b27cb](https://github.com/mozilla/addons-frontend/commit/57b27cb31c2c2712059c189e1e492ed83910da61)) into [master](https://codecov.io/gh/mozilla/addons-frontend/commit/c58c53ba1e0117b6a28381d8837e86826ebb1acb?el=desc) ([c58c53b](https://github.com/mozilla/addons-frontend/commit/c58c53ba1e0117b6a28381d8837e86826ebb1acb)) will **decrease** coverage by `0.37%`.
> > The diff coverage is `63.79%`.
> 
> mm this is weird. The diff coverage remains 63% but when looking in codecov, it seems a bit better than that. We definitely want 100% diff coverage.
>

I agree this is odd. I do know that in the past I have seen cases where a codecov report is generated when the PR is first created, and doesn't get updated as the PR is updated. If you follow the link to codecov, the current report there looks good: https://app.codecov.io/gh/mozilla/addons-frontend/compare/10065/diff, but I'm not sure how accurate that is.

One thing we could do would be to create a new PR using the same commits and see what codecov generates for that.

@xlisachan It's also suggested you rebase these changes on top of the current master, so maybe you can do that and open a new PR. Let's keep this one open too, for now, so we can track the comments. If you have trouble rebasing this because of the large number of commits, you could squash them all into a single commit first.

> I'd suggest you rebase your PR, too.

See above 😉

xlisachan commented:
Thanks for the feedback! Per requested, I have created a new PR (see above) with the same commits to see what codecov generates. Within that pull request, I made updates to the following:
- `tests/unit/helpers.js`
- `tests/unit/amo/sagas/test_home.js`
- `src/amo/components/HomepageShelves/index.js`

Please review that pull request when you have a moment. Thanks!
